lib/media/media_source_engine.js

```diff
@@ -1513,6 +1554,181 @@ shaka.media.MediaSourceEngine = class {
+      return;
+    }
+    shaka.log.debug(
+        `Change Type: ${this.sourceBufferTypes_[contentType]} -> ${mimeType}`);
+    if (shaka.media.Capabilities.isChangeTypeSupported()) {
+      if (this.transmuxers_[contentType]) {
+        this.transmuxers_[contentType].destroy();
+      }
+      if (transmuxer) {
+        this.transmuxers_[contentType] = transmuxer;
+      }
+      const type = mimeType + this.config_.sourceBufferExtraFeatures;
+      this.sourceBuffers_[contentType].changeType(type);
+      this.sourceBufferTypes_[contentType] = mimeType;
+    } else {
+      shaka.log.debug('Change Type not supported');
+    }
+
+    // Fake an 'updateend' event to resolve the operation.
+    this.onUpdateEnd_(contentType);
+  }
+
+  /**
+   * Enqueue an operation to prepare the SourceBuffer to parse a potentially new
+   * type or codec.
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @param {string} mimeType
+   * @param {?shaka.extern.Transmuxer} transmuxer
+   * @return {!Promise}
+   */
+  changeType(contentType, mimeType, transmuxer) {
+    return this.enqueueOperation_(
+        contentType,
+        () => this.change_(contentType, mimeType, transmuxer));
+  }
+
+  /**
+   * Resets the MediaSource and re-adds source buffers due to codec mismatch
+   *
+   * @param {!Map.<shaka.util.ManifestParserUtils.ContentType,
+   *               shaka.extern.Stream>} streamsByType
+   * @private
+   */
+  async reset_(streamsByType) {
+    this.reloadingMediaSource_ = true;
+    const currentTime = this.video_.currentTime;
+    try {
+      this.eventManager_.removeAll();
+      this.sourceBuffers_ = {};
+
+      const cleanup = [];
+      for (const contentType in this.transmuxers_) {
+        cleanup.push(this.transmuxers_[contentType].destroy());
+      }
+      await Promise.all(cleanup);
+      this.transmuxers_ = {};
+
+      const previousDuration = this.mediaSource_.duration;
+      this.mediaSourceOpen_ = new shaka.util.PublicPromise();
+      this.mediaSource_ = this.createMediaSource(this.mediaSourceOpen_);
+      await this.mediaSourceOpen_;
+      this.mediaSource_.duration = previousDuration;
+
+      const sourceBufferAdded = new shaka.util.PublicPromise();
+      const sourceBuffers =
+        /** @type {EventTarget} */(this.mediaSource_.sourceBuffers);
+
+      const totalOfBuffers = streamsByType.size;
+      let numberOfSourceBufferAdded = 0;
+      this.eventManager_.listen(sourceBuffers, 'addsourcebuffer', (event) => {
+        numberOfSourceBufferAdded++;
+        if (numberOfSourceBufferAdded === totalOfBuffers) {
+          sourceBufferAdded.resolve();
+        }
+      });
+
+      for (const contentType of streamsByType.keys()) {
+        const stream = streamsByType.get(contentType);
+        this.initSourceBuffer_(contentType, stream);
+      }
+
+      // Fake a seek to catchup the playhead.
+      this.video_.currentTime = currentTime;
+
+      await sourceBufferAdded;
```

Comment:
It's a good question, I've always seen that addsourcebuffer is fired, but... I could be wrong.

Reply:
Hm. I'm a little worried, because the `reloadingMediaSource_` flag being true prevents us from detecting the media source engine as ended. So if one of the source buffers does error during this process we won't shut down properly, right?

Maybe you should add event listeners to the buffers for `error`, and turn this await into a `Promise.race`.

Reply:
Ok, we discussed this over messaging and decided that just catching the errors is probably enough.